File: kotlin/src/test/kotlin/com/automation/agent/obs/MiddlewareTest.kt
/*
 * Tests for the HTTP tracing interceptor. Deterministic: no live network beyond Ktor's in-memory
 * test host, no LLM. Assertions are on the server span's name / kind / status / parenting, and on the
 * health-probe exclusion and the flush-before-return guarantee.
 */
package com.automation.agent.obs

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.shouldBe
import io.opentelemetry.api.GlobalOpenTelemetry
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.context.Context
import io.ktor.client.request.header
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.http.HttpStatusCode
import io.ktor.server.response.respond
import io.ktor.server.response.respondText
import io.ktor.server.routing.get
import io.ktor.server.routing.post
import io.ktor.server.routing.routing
import io.ktor.server.testing.testApplication

class MiddlewareTest : BehaviorSpec({
    Given("a traced request") {
        When("it is handled") {
            Then("one server span is exported with the method-path name and status attribute") {
                withRecording { exporter ->
                    testApplication {
                        application {
                            installObsTracing()
                            routing { post("/webhooks/lint") { call.respond(HttpStatusCode.Accepted) } }
                        }
                        client.post("/webhooks/lint").status shouldBe HttpStatusCode.Accepted
                    }
                    flush()

                    val spans = exporter.finishedSpanItems
                    spans shouldHaveSize 1
                    spans[0].name shouldBe "POST /webhooks/lint"
                    spans[0].kind shouldBe SpanKind.SERVER
                    spans[0].attributes.get(AttributeKey.longKey("http.response.status_code")) shouldBe 202L
                }
            }
        }
    }

    Given("the health probe") {
        When("it is polled") {
            Then("it is excluded from tracing") {
                withRecording { exporter ->
                    testApplication {
                        application {
                            installObsTracing()
                            routing { get("/healthz") { call.respondText("ok") } }
                        }
                        client.get("/healthz").status shouldBe HttpStatusCode.OK
                    }
                    flush()
                    exporter.finishedSpanItems shouldHaveSize 0
                }
            }
        }
    }

    Given("a buffered span and the health probe") {
        When("the probe is handled") {
            Then("it does not flush other requests' buffered spans, but a traced request does") {
                withRecording { exporter ->
                    // Buffer a span (as an in-flight agent run would).
                    GlobalOpenTelemetry.getTracer("obs-test")
                        .spanBuilder("invoke_agent automation_agent").startSpan().end()
                    testApplication {
                        application {
                            installObsTracing()
                            routing {
                                get("/healthz") { call.respondText("ok") }
                                post("/webhooks/lint") { call.respond(HttpStatusCode.Accepted) }
                            }
                        }
                        client.get("/healthz")
                        // The probe left the buffered span un-exported (no span, no flush).
                        exporter.finishedSpanItems shouldHaveSize 0
                        // A traced request flushes them (its own server span plus the buffered one).
                        client.post("/webhooks/lint")
                    }
                    flush()
                    exporter.finishedSpanItems.size shouldBeGreaterThan 0
                }
            }
        }
    }

    Given("a request that returns 5xx") {
        When("it is handled") {
            Then("the server span records an error status") {
                withRecording { exporter ->
                    testApplication {
                        application {
                            installObsTracing()
                            routing { post("/webhooks/lint") { call.respond(HttpStatusCode.InternalServerError, "boom") } }
                        }
                        client.post("/webhooks/lint").status shouldBe HttpStatusCode.InternalServerError
                    }
                    flush()
                    val span = exporter.finishedSpanItems.first { it.name == "POST /webhooks/lint" }
                    span.status.statusCode shouldBe StatusCode.ERROR
                }
            }
        }
    }

    Given("a handler that throws") {
        When("the exception propagates") {
            Then("the span records the exception and keeps its message (not overwritten by the 5xx branch)") {
                withRecording { exporter ->
                    testApplication {
                        application {
                            installObsTracing()
                            routing { post("/webhooks/lint") { throw RuntimeException("boom") } }
                        }
                        // The engine turns the unhandled throw into a 500.
                        client.post("/webhooks/lint").status shouldBe HttpStatusCode.InternalServerError
                    }
                    flush()
                    val span = exporter.finishedSpanItems.first { it.name == "POST /webhooks/lint" }
                    span.status.statusCode shouldBe StatusCode.ERROR
                    // The exception message survives to the top-level status.
                    span.status.description shouldBe "boom"
                    span.events.any { it.name == "exception" } shouldBe true
                }
            }
        }
    }

    Given("a task carrying an inbound traceparent") {
        When("the dispatch request is handled") {
            Then("the server span continues the ingress trace as its child") {
                withRecording { exporter ->
                    // Model a task carrying a traceparent injected by the transport on enqueue.
                    val ingress = GlobalOpenTelemetry.getTracer("obs-test").spanBuilder("ingress").startSpan()
                    val carrier = inject(mutableMapOf(), Context.current().with(ingress))
                    ingress.end()

                    testApplication {
                        application {
                            installObsTracing()
                            routing { post("/internal/dispatch") { call.respond(HttpStatusCode.OK, "ok") } }
                        }
                        client.post("/internal/dispatch") { header("traceparent", carrier.getValue("traceparent")) }
                    }
                    flush()

                    val dispatch = exporter.finishedSpanItems.first { it.name == "POST /internal/dispatch" }
                    dispatch.traceId shouldBe ingress.spanContext.traceId
                    dispatch.parentSpanContext.spanId shouldBe ingress.spanContext.spanId
                }
            }
        }
    }

    Given("tracing disabled") {
        When("a request is handled") {
            Then("the interceptor is a true no-op and the request still completes") {
                GlobalOpenTelemetry.resetForTest()
                var called = false
                testApplication {
                    application {
                        installObsTracing()
                        routing { post("/webhooks/lint") { called = true; call.respond(HttpStatusCode.Accepted) } }
                    }
                    client.post("/webhooks/lint").status shouldBe HttpStatusCode.Accepted
                }
                called shouldBe true
            }
        }
    }
})

File: javascript/src/obs/middleware.test.ts
/**
 * Tests for the HTTP tracing middleware. Deterministic: no live network, no LLM. The middleware is
 * driven directly with mock request/response objects so the response lifecycle (and thus the
 * end-of-request flush) is controlled explicitly.
 */
import { EventEmitter } from 'node:events';
import { type NextFunction, type Request, type Response } from 'express';

import { SpanKind, SpanStatusCode, context, propagation, trace } from '@opentelemetry/api';
import { type ReadableSpan, InMemorySpanExporter } from '@opentelemetry/sdk-trace-base';
import { afterEach, describe, expect, it } from 'vitest';

import { HEALTH_PATH, httpMiddleware } from './middleware';
import { TracesExporter, flush, install } from './obs';
import { inject } from './propagation';

function resetOtel(): void {
  trace.disable();
  propagation.disable();
  context.disable();
}

function recording(): { exporter: InMemorySpanExporter; shutdown: () => Promise<void> } {
  const exporter = new InMemorySpanExporter();
  const shutdown = install(exporter, { exporter: TracesExporter.Console, serviceName: 'automation-agent' });
  return { exporter, shutdown };
}

function mockReq(method: string, path: string, headers: Record<string, string> = {}): Request {
  return { method, path, headers } as unknown as Request;
}

/** A response stub that carries a status code and emits lifecycle events like the real one. */
function mockRes(statusCode = 202): Response & EventEmitter {
  const res = new EventEmitter() as Response & EventEmitter;
  res.statusCode = statusCode;
  return res;
}

/** Emit an ended agent span so the batch buffer has something to (not) flush. */
function bufferSpan(): void {
  trace.getTracer('obs-test').startSpan('invoke_agent automation_agent').end();
}

function findSpan(exporter: InMemorySpanExporter, name: string): ReadableSpan {
  const span = exporter.getFinishedSpans().find((s) => s.name === name);
  if (!span) {
    throw new Error(`span not found: ${name}`);
  }
  return span;
}

const noopNext: NextFunction = () => {};

describe('obs httpMiddleware', () => {
  afterEach(resetOtel);

  it('creates one server span per request without altering the response', async () => {
    const { exporter, shutdown } = recording();
    try {
      const mw = httpMiddleware();
      const res = mockRes(202);
      let called = 0;
      mw(mockReq('POST', '/webhooks/lint'), res, () => {
        called += 1;
      });
      res.emit('finish');
      await flush();

      expect(called).toBe(1);
      const spans = exporter.getFinishedSpans();
      expect(spans).toHaveLength(1);
      expect(spans[0]!.name).toBe('POST /webhooks/lint');
      expect(spans[0]!.kind).toBe(SpanKind.SERVER);
      expect(spans[0]!.attributes['http.response.status_code']).toBe(202);
    } finally {
      await shutdown();
    }
  });

  it('excludes the health probe from tracing', async () => {
    const { exporter, shutdown } = recording();
    try {
      let called = 0;
      httpMiddleware()(mockReq('GET', HEALTH_PATH), mockRes(200), () => {
        called += 1;
      });
      await flush();
      expect(called).toBe(1);
      expect(exporter.getFinishedSpans()).toHaveLength(0);
    } finally {
      await shutdown();
    }
  });

  it('does not flush on the health probe', async () => {
    const { exporter, shutdown } = recording();
    try {
      bufferSpan();
      const mw = httpMiddleware();

      // The health probe leaves the buffered span un-exported (no span, no flush).
      mw(mockReq('GET', HEALTH_PATH), mockRes(200), noopNext);
      expect(exporter.getFinishedSpans()).toHaveLength(0);

      // A traced request flushes them (its own server span plus the buffered one).
      const res = mockRes(202);
      mw(mockReq('POST', '/webhooks/lint'), res, noopNext);
      res.emit('finish');
      await flush();
      expect(exporter.getFinishedSpans().length).toBeGreaterThan(0);
    } finally {
      await shutdown();
    }
  });

  it('records an error status on a 5xx response', async () => {
    const { exporter, shutdown } = recording();
    try {
      const res = mockRes(500);
      httpMiddleware()(mockReq('POST', '/webhooks/lint'), res, noopNext);
      res.emit('finish');
      await flush();
      expect(findSpan(exporter, 'POST /webhooks/lint').status.code).toBe(SpanStatusCode.ERROR);
    } finally {
      await shutdown();
    }
  });

  it('records an exception when the chain throws synchronously', async () => {
    const { exporter, shutdown } = recording();
    try {
      const boom = new Error('boom');
      // The framework's error handling turns the throw into a 500 and completes the response.
      const res = mockRes(500);
      expect(() =>
        httpMiddleware()(mockReq('POST', '/webhooks/lint'), res, () => {
          throw boom;
        }),
      ).toThrow('boom');
      // The span records the exception on throw but is completed by the response event (not in the
      // catch), so the recorded status reflects the real response the framework ultimately sends.
      res.emit('close');
      await flush();

      const span = findSpan(exporter, 'POST /webhooks/lint');
      expect(span.status.code).toBe(SpanStatusCode.ERROR);
      expect(span.events.some((e) => e.name === 'exception')).toBe(true);
    } finally {
      await shutdown();
    }
  });

  it('continues an incoming trace from a task traceparent header', async () => {
    const { exporter, shutdown } = recording();
    try {
      // Model a task carrying a traceparent injected by the transport on enqueue.
      const ingress = trace.getTracer('obs-test').startSpan('ingress');
      const carrier = inject({}, trace.setSpan(context.active(), ingress));
      ingress.end();

      const res = mockRes(200);
      httpMiddleware()(mockReq('POST', '/internal/dispatch', carrier), res, noopNext);
      res.emit('finish');
      await flush();

      const dispatch = findSpan(exporter, 'POST /internal/dispatch');
      expect(dispatch.spanContext().traceId).toBe(ingress.spanContext().traceId);
      expect(dispatch.parentSpanContext?.spanId).toBe(ingress.spanContext().spanId);
    } finally {
      await shutdown();
    }
  });

  it('is a true no-op when tracing is disabled', () => {
    // No provider installed: the middleware calls next and attaches no listeners, so a request
    // with no lifecycle events still completes and nothing is traced.
    let called = 0;
    httpMiddleware()(mockReq('POST', '/webhooks/lint'), mockRes(202), () => {
      called += 1;
    });
    expect(called).toBe(1);
  });
});
